src/bitwidth.rs

```diff
@@ -11,68 +14,129 @@ use crate::{
+///     // the function, and the function will handle both `BitWidth` invariant checking
+///     // and its own invariants.
+///     x.sign_extend(100)?;
+///
+///     Ok(x)
+/// }
+///
+/// example_function(64).unwrap();
+/// ```
 #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct BitWidth(usize);
+pub struct BitWidth(NonZeroUsize);
 
-//  ===========================================================================
-///  Constructors
-/// ===========================================================================
-impl BitWidth {
-    /// Creates a `BitWidth` that represents a bit-width of `1` bit.
-    #[inline]
-    pub fn w1() -> Self {
-        BitWidth(1)
-    }
-
-    /// Creates a `BitWidth` that represents a bit-width of `8` bits.
-    #[inline]
-    pub fn w8() -> Self {
-        BitWidth(8)
-    }
-
-    /// Creates a `BitWidth` that represents a bit-width of `16` bits.
-    #[inline]
-    pub fn w16() -> Self {
-        BitWidth(16)
-    }
-
-    /// Creates a `BitWidth` that represents a bit-width of `32` bits.
-    #[inline]
-    pub fn w32() -> Self {
-        BitWidth(32)
+impl From<NonZeroUsize> for BitWidth {
+    /// Creates a `BitWidth` from the given `NonZeroUsize`.
+    fn from(width: NonZeroUsize) -> Self {
+        BitWidth(width)
     }
+}
 
-    /// Creates a `BitWidth` that represents a bit-width of `64` bits.
-    #[inline]
-    pub fn w64() -> Self {
-        BitWidth(64)
-    }
-
-    /// Creates a `BitWidth` that represents a bit-width of `128` bits.
-    #[inline]
-    pub fn w128() -> Self {
-        BitWidth(128)
-    }
+impl TryFrom<usize> for BitWidth {
+    type Error = crate::Error;
 
     /// Creates a `BitWidth` from the given `usize`.
     ///
     /// # Errors
     ///
     /// - If the given `width` is equal to zero.
-    pub fn new(width: usize) -> Result<Self> {
-        if width == 0 {
-            return Err(Error::invalid_zero_bitwidth())
+    fn try_from(width: usize) -> Result<BitWidth> {
+        match NonZeroUsize::new(width) {
+            Some(bitwidth) => Ok(BitWidth(bitwidth)),
+            None => Err(crate::Error::invalid_zero_bitwidth()),
+        }
+    }
+}
+
+/// Convenience free function for converting a `usize` to a `BitWidth`.
+///
+/// # Panics
+///
+/// If `width == 0`, this function will panic
+pub const fn bw(width: usize) -> BitWidth {
+    // Ideally, the code here would be `width.try_into().ok().expect("tried to
+    // ...");`, but we want this to be a `const` function.
+    match NonZeroUsize::new(width) {
+        None => {
+            panic!(
+                "Tried to construct an invalid BitWidth of 0 using the `apint::bw` \
+                 function"
+            )
         }
-        Ok(BitWidth(width))
+        Some(n) => BitWidth(n),
     }
+}
+
+impl BitWidth {
+    // TODO: change to using `NonZeroUsize::new()` as soon as `unwrap`ing can be
+    // done.
```

Comment:
TODO: please remove or convert into issue

Reply:
I will just remove these constants altogether since they only helped in a few places